base_layer/p2p/src/services/executor.rs

```diff
@@ -0,0 +1,221 @@
+
+const LOG_TARGET: &'static str = "base_layer::p2p::services";
+
+pub enum ServiceControlMessage {
+    Shutdown,
+}
+
+pub struct ServiceExecutor {
+    thread_pool: ThreadPool,
+    senders: Vec<Sender<ServiceControlMessage>>,
+    comms_services: Arc<CommsServices<TariMessageType>>,
+}
+
+impl ServiceExecutor {
+    pub fn execute(comms_services: Arc<CommsServices<TariMessageType>>, registry: ServiceRegistry) -> Self {
+        let thread_pool = threadpool::Builder::new()
+            .thread_name("DomainServices".to_string())
+            .num_threads(registry.num_services())
+            .thread_stack_size(1_000_000)
+            .build();
+
+        let mut senders = Vec::new();
+
+        for mut service in registry.services.into_iter() {
+            let (sender, receiver) = channel();
+            senders.push(sender);
+
+            let service_context = ServiceContext {
+                comms_services: comms_services.clone(),
+                receiver,
+            };
+
+            thread_pool.execute(move || {
+                debug!(target: LOG_TARGET, "Executing service {}", service.get_name());
+                service.execute(service_context);
+            });
+        }
+
+        Self {
+            thread_pool,
+            senders,
+            comms_services,
+        }
+    }
+
+    pub fn shutdown(&self) -> Result<(), ServiceError> {
+        let mut failed = false;
+        for sender in &self.senders {
+            if sender.send(ServiceControlMessage::Shutdown).is_err() {
+                failed = true;
+            }
+        }
+
+        // TODO: Wait for services to exit and then shutdown the comms
+        //        self.comms_services
+        //            .shutdown()
+        //            .map_err(ServiceError::CommsServicesError)?;
+
+        if failed {
+            Err(ServiceError::ShutdownSendFailed)
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn join_timeout(self, timeout: Duration) -> Result<(), ServiceError> {
+        let (tx, rx) = channel();
+        thread::spawn(move || {
+            self.thread_pool.join();
+            let _ = tx.send(());
+        });
+
+        rx.recv_timeout(timeout).map_err(|_| ServiceError::JoinTimedOut)
+    }
+}
+
+pub struct ServiceContext {
+    comms_services: Arc<CommsServices<TariMessageType>>,
+    receiver: Receiver<ServiceControlMessage>,
+}
+
+impl ServiceContext {
+    pub fn get_control_message(&self, timeout: Duration) -> Option<ServiceControlMessage> {
+        match self.receiver.recv_timeout(timeout) {
+            Ok(msg) => Some(msg),
+            // Sender has disconnected (dropped) so return a shutdown signal
+            // This should never happen in normal operation
+            Err(RecvTimeoutError::Disconnected) => Some(ServiceControlMessage::Shutdown),
+            Err(RecvTimeoutError::Timeout) => None,
+        }
+    }
+
+    pub fn get_outbound_message_service(&self) -> &Arc<OutboundMessageService> {
```

Comment:
Makes sense in this case - once something is in an Arc you can't get it out (can you only try with `Arc::try_unwrap`) - we have it in an Arc because we don't want to clone the rather fat object each time - here I don't want to create unnecessary atomic references (one Arc needed per service/thread), so I figured the ServiceContext can hold onto this. Happy to change it though. IMO the issue I think is the OMS - we need to put that behind a thinner object or a trait or something - OMS also makes services hard to test.

Reply:
Then I believe you really want to return `&OMS`

Reply:
Agreed - they can't 'leak'  so it's just any overhead in cloning them (which I believe is small) - made some changes here